common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationFactory.java

```diff
@@ -0,0 +1,171 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.catalyst.util;
+
+import java.util.*;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+import com.ibm.icu.util.ULocale;
+import com.ibm.icu.text.Collator;
+
+import org.apache.spark.SparkIllegalArgumentException;
+import org.apache.spark.unsafe.types.UTF8String;
+
+/**
+ * Static entry point for collation aware string functions.
+ * Provides functionality to the UTF8String object which respects defined collation settings.
+ */
+public final class CollationFactory {
+  /**
+   * Entry encapsulating all information about a collation.
+   */
+  public static class CollationInfo {
+    public final String collationName;
+    public final Collator collator;
+    public final Comparator<UTF8String> comparator;
+
+    /**
+     * Version of the collation. This is the version of the ICU library used to create the collator.
+     * For non-ICU collations (e.g. UTF8 Binary) the version is set to "1.0".
+     */
+    public final String version;
+
+    /**
+     * Collation sensitive hash function. Output for two UTF8Strings will be the same if they are
+     * equal according to the collation.
+     */
+    public final Function<UTF8String, Integer> hashFunction;
+
+    /**
+     * Potentially faster way than using comparator to compare two UTF8Strings for equality.
+     * Falls back to binary comparison if the collation is binary.
+     */
+    public final BiFunction<UTF8String, UTF8String, Boolean> equalsFunction;
+
+    /**
+     * Binary collation implies that UTF8Strings are considered equal only if they are
+     * byte for byte equal. All accent or case-insensitive collations are considered non-binary.
+     */
+    public final boolean isBinaryCollation;
+
+    public CollationInfo(
+      String collationName,
+      Collator collator,
+      Comparator<UTF8String> comparator,
+      String version,
+      Function<UTF8String, Integer> hashFunction,
+      boolean isBinaryCollation) {
+      this.collationName = collationName;
+      this.collator = collator;
+      this.comparator = comparator;
+      this.version = version;
+      this.hashFunction = hashFunction;
+      this.isBinaryCollation = isBinaryCollation;
+
+      if (isBinaryCollation) {
+        this.equalsFunction = UTF8String::binaryEquals;
```

Comment:
shouldn't it be `UTF8String::equals`?

Reply:
With latest changes that keep UTF8String intact this conversation is no longer relevant. I removed binaryEquals from this PR.